### PR TITLE
add cross domain scoping

### PR DESCRIPTION
**Pull request by auhlig**
As described in https://github.com/improbable-eng/thanos/issues/1117

### CODE REVIEW

File: pkg/objstore/swift/swift.go [hunk 1]
```diff
@@ -12,6 +12,7 @@ import (
 	"time"
 
 	"github.com/improbable-eng/thanos/pkg/objstore"
+	yaml "gopkg.in/yaml.v2"
```
Comment (povilasv) [inline pkg/objstore/swift/swift.go]: please remove the blank line after this one

File: pkg/objstore/swift/swift.go [hunk 2]
```diff
@@ -65,6 +76,36 @@ func NewContainer(logger log.Logger, conf []byte) (*Container, error) {
...
+		authOpts.TenantID = sc.ProjectID
+	}
+	if sc.ProjectName != "" {
+		authOpts.TenantName = sc.ProjectName
+	}
+
+	// Support for cross domain scoping (user in different domain than project)
+	// If a userDomainName or userDomainID is given, a user is scoped to this domain
+	// and the tenant (aka project) is expected to be in the domain given by domainName or domainID.
+	if sc.UserDomainName != "" {
```
Comment (povilasv) [inline pkg/objstore/swift/swift.go]: could we use switch here instead

```
switch {
case sc.UserDomaiName != "": 
...
case sc.UserDomainID != "":

}
```

File: pkg/objstore/swift/swift.go [hunk 3]
```diff
@@ -65,6 +76,36 @@ func NewContainer(logger log.Logger, conf []byte) (*Container, error) {
...
+			DomainName: sc.DomainName,
+		}
+	} else if sc.UserDomainID != "" {
+		authOpts.DomainID = sc.UserDomainID
+		authOpts.Scope = &gophercloud.AuthScope{
+			DomainID: sc.DomainId,
+		}
+	}
+	if authOpts.Scope != nil {
+		if sc.TenantName != "" {
```
Comment (povilasv) [inline pkg/objstore/swift/swift.go]: could we use switch here.

File: pkg/objstore/swift/swift.go (hunk 1 above)
Comment (auhlig) [reply pkg/objstore/swift/swift.go]: ran `goimports` again

File: pkg/objstore/swift/swift.go (hunk 2 above)
Comment (auhlig) [reply pkg/objstore/swift/swift.go]: right. done

File: pkg/objstore/swift/swift.go (hunk 3 above)
Comment (auhlig) [reply pkg/objstore/swift/swift.go]: done

File: pkg/objstore/swift/swift.go (hunk 1 above)
Comment (povilasv) [reply pkg/objstore/swift/swift.go]: goimports doesn't solve the blank line issue, please remove the blank line after this

**Comment by povilasv** [comment]
@auhlig you just commented, but didn't actually push the changes

File: pkg/objstore/swift/swift.go (hunk 1 above)
Comment (auhlig) [reply pkg/objstore/swift/swift.go]: i meant removed blank line and did the goimports again

**Comment by auhlig** [comment]
@povilasv github was down this morning -.-

File: pkg/objstore/swift/swift.go [hunk 4]
```diff
@@ -58,13 +69,57 @@ func NewContainer(logger log.Logger, conf []byte) (*Container, error) {
...
+	// In Identity v3 the term `project` is used instead.
+	switch {
+	case sc.TenantName != "":
+		level.Warn(logger).Log("msg", "usage of tenant_name is deprecated. use project_name instead")
+		authOpts.TenantName = sc.TenantName
+	case sc.TenantID != "":
+		level.Warn(logger).Log("msg", "usage of tenant_id is deprecated. use project_id instead")
+		authOpts.TenantID = sc.TenantID
+	case sc.ProjectName != "" || sc.ProjectID != "":
+		level.Info(logger).Log("msg", "using project_id or project_name")
```
Comment (povilasv) [inline pkg/objstore/swift/swift.go]: I think we should remove this, it isn't a good message and IMO doesn't add any value

File: pkg/objstore/swift/swift.go [hunk 5]
```diff
@@ -58,13 +69,57 @@ func NewContainer(logger log.Logger, conf []byte) (*Container, error) {
...
+		authOpts.Scope = &gophercloud.AuthScope{
+			DomainName: sc.ProjectDomainName,
+		}
+	case sc.ProjectDomainID != "":
+		authOpts.Scope = &gophercloud.AuthScope{
+			DomainID: sc.ProjectDomainID,
+		}
+	}
+	if authOpts.Scope != nil {
+		level.Info(logger).Log("msg", "scoping to domain, project")
```
Comment (povilasv) [inline pkg/objstore/swift/swift.go]: same here let's not log this message.

File: pkg/objstore/swift/swift.go [hunk 6]
```diff
@@ -58,13 +69,57 @@ func NewContainer(logger log.Logger, conf []byte) (*Container, error) {
...
+		TenantID:         sc.ProjectID,
+		TenantName:       sc.ProjectName,
 
 		// Allow Gophercloud to re-authenticate automatically.
 		AllowReauth: true,
 	}
 
+	// The term `tenant` is used in the deprecated OpenStack Identity v2.
+	// In Identity v3 the term `project` is used instead.
+	switch {
```
Comment (povilasv) [inline pkg/objstore/swift/swift.go]: could you also add a couple of examples how or when to configure project id vs project name in the `docs/storage.md` file

File: pkg/objstore/swift/swift.go (hunk 4 above)
Comment (auhlig) [reply pkg/objstore/swift/swift.go]: right. removed

File: pkg/objstore/swift/swift.go (hunk 5 above)
Comment (auhlig) [reply pkg/objstore/swift/swift.go]: removed

File: pkg/objstore/swift/swift.go (hunk 6 above)
Comment (auhlig) [reply pkg/objstore/swift/swift.go]: updated docs. added link to official documentation with examples

**Comment by auhlig** [comment]
LGTY @povilasv?

**Comment by povilasv** [comment]
@auhlig Sort of yes. Did you look at the tests? Would be amazing if we could add some proper testing here, so that when we change code we don't break things.

File: pkg/objstore/swift/swift.go [hunk 7]
```diff
@@ -11,32 +11,43 @@ import (
...
+	DomainId          string `yaml:"domain_id"`
+	DomainName        string `yaml:"domain_name"`
+	ProjectID         string `yaml:"project_id"`
+	ProjectName       string `yaml:"project_name"`
+	ProjectDomainID   string `yaml:"project_domain_id"`
+	ProjectDomainName string `yaml:"project_domain_name"`
+	RegionName        string `yaml:"region_name"`
+	ContainerName     string `yaml:"container_name"`
+
+	// Deprecated: Please use `project_id` instead.
```
Comment (bwplotka) [inline pkg/objstore/swift/swift.go]: I would maybe kill this and use `yaml.UmarshalStrict` instead? (: We already documented the change so let's fail fast on umarshalling! (:

File: pkg/objstore/swift/swift.go [hunk 8]
```diff
@@ -58,13 +69,54 @@ func NewContainer(logger log.Logger, conf []byte) (*Container, error) {
...
+		TenantName:       sc.ProjectName,
 
 		// Allow Gophercloud to re-authenticate automatically.
 		AllowReauth: true,
 	}
 
+	// The term `tenant` is used in the deprecated OpenStack Identity v2.
+	// In Identity v3 the term `project` is used instead.
+	switch {
+	case sc.TenantName != "":
```
Comment (bwplotka) [inline pkg/objstore/swift/swift.go]: let's use `yaml.UnmarshalStrict` instead?

File: pkg/objstore/swift/swift.go [hunk 9]
```diff
@@ -180,13 +232,19 @@ func (c *Container) deleteContainer(name string) error {
...
+		Password:          os.Getenv("OS_PASSWORD"),
+		RegionName:        os.Getenv("OS_REGION_NAME"),
+		ContainerName:     os.Getenv("OS_CONTAINER_NAME"),
+		ProjectID:         os.Getenv("OS_PROJECT_ID"),
+		ProjectName:       os.Getenv("OS_PROJECT_NAME"),
+		UserDomainID:      os.Getenv("OS_USER_DOMAIN_ID"),
+		UserDomainName:    os.Getenv("OS_USER_DOMAIN_NAME"),
+		ProjectDomainID:   os.Getenv("OS_PROJET_DOMAIN_ID"),
+		ProjectDomainName: os.Getenv("OS_PROJECT_DOMAIN_NAME"),
+		TenantID:          os.Getenv("OS_TENANT_ID"),
```
Comment (bwplotka) [inline pkg/objstore/swift/swift.go]: Kill this

**Comment by auhlig** [comment]
Made this a breaking change by dropping `tenant_id`, `tenant_name`. 
Extracted logic to `parseConfig`, `authOptsFromConfig` for easier unit testing.
Added unit tests.

LGTY @povilasv @bwplotka

**Comment by auhlig** [comment]
Netlify doesn't seem to be in the mood to run the tests. CircleCI passed however.

File: pkg/objstore/swift/swift.go (hunk 9 above)
Comment (auhlig) [reply pkg/objstore/swift/swift.go]: killed

File: pkg/objstore/swift/swift.go (hunk 8 above)
Comment (auhlig) [reply pkg/objstore/swift/swift.go]: done

File: pkg/objstore/swift/swift.go (hunk 7 above)
Comment (auhlig) [reply pkg/objstore/swift/swift.go]: sounds good. done

**Comment by auhlig** [comment]
squashed into single commit and force pushed.
giving netlify another chance

**Comment by auhlig** [comment]
Updated the changelog after some merge conflicts. Can this be merged @povilasv @bwplotka?

**Comment by GiedriusS** [comment]
@auhlig so essentially with this PR we are getting rid of the support of the V2 API? I'm a bit anxious about this since, for example, some companies might still be running old versions of that API (*cough* not gonna point any fingers). Would it be hard to add some kind of backward compatibility?

**Comment by auhlig** [comment]
@GiedriusS This PR does **not** drop Identity v2 support. It's still possible to authenticate using `project_id` or `project_name` instead of the deprecated terms `tenant_id`, `tenant_name`.
We only fix the terminology ( tenant became project with Identity v3) and add support for for authentication if a  user is in a different domain than the project.
In summary, the only breaking change is the terminology.

**Comment by povilasv** [comment]
Ok so this does include a breaking change
`tenantID`, `tenantName` -> `projectID`, `projectName`.

I think in order not to break users, just fill tenantID to projectID and tenantName to projectName.

We try to avoid breaking changes when possible and in this case it's definitely doable

**Comment by bwplotka** [comment]
@povilasv I think breaking here is better that having inconsistent configuration -> users will be immdiately notified by verbose error message.

**Comment by bwplotka** [comment]
Waiting for @povilasv final LGTM to merge (:

**Comment by auhlig** [comment]
Thank you guys 🎉

**Pull request merged**